### PR TITLE
Add FinalizerQueue to ObjectFactory

### PR DESCRIPTION
**Pull request by projedi**

### CODE REVIEW

File: runtime/src/mm/cpp/ObjectFactory.hpp [hunk 1]
```diff
@@ -230,18 +284,19 @@ class ObjectFactoryStorage : private Pinned {
...
-    Node* EraseUnsafe(Node* previousNode) noexcept {
+    std::pair<unique_ptr<Node>, Node*> ExtractUnsafe(Node* previousNode) noexcept {
         RuntimeAssert(root_ != nullptr, "Must not be empty");
         AssertCorrectUnsafe();
 
         if (previousNode == nullptr) {
+            auto node = std::move(root_);
             // Deleting the root.
-            root_ = std::move(root_->next_);
+            root_ = std::move(node->next_);
```
Comment (SvyatoslavScherbina) [inline runtime/src/mm/cpp/ObjectFactory.hpp:294]: The comment above is probably obsolete.

File: runtime/src/mm/cpp/ObjectFactoryTest.cpp [hunk 2]
```diff
@@ -317,6 +329,173 @@ TEST(ObjectFactoryStorageTest, EraseTheOnlyElement) {
...
+    Consumer<ObjectFactoryStorageRegular> consumer;
+
+    producer.Insert<int>(1);
+
+    producer.Publish();
+
+    {
+        auto iter = storage.Iter();
+        auto it = iter.begin();
+        iter.MoveAndAdvance(consumer, it);
```
Comment (SvyatoslavScherbina) [inline runtime/src/mm/cpp/ObjectFactoryTest.cpp:456]: Checking that `it == iter.end()` here makes sense.

**Review by SvyatoslavScherbina — APPROVED** [review]

File: runtime/src/mm/cpp/ObjectFactory.hpp (hunk 1 above)
Comment (projedi) [reply runtime/src/mm/cpp/ObjectFactory.hpp:294]: Yep, thanks.

File: runtime/src/mm/cpp/ObjectFactoryTest.cpp (hunk 2 above)
Comment (projedi) [reply runtime/src/mm/cpp/ObjectFactoryTest.cpp:457]: Good point, also fixed the related test with `EraseAndAdvance`.

**Review by SvyatoslavScherbina — APPROVED** [review]

File: runtime/src/mm/cpp/ObjectFactory.hpp [hunk 3]
```diff
@@ -206,14 +206,68 @@ class ObjectFactoryStorage : private Pinned {
...
+        Consumer(Consumer&&) noexcept = default;
+        Consumer& operator=(Consumer&&) noexcept = default;
+
+        ~Consumer() {
+            // Make sure not to blow up the stack by nested `~Node` calls.
+            for (auto node = std::move(root_); node != nullptr; node = std::move(node->next_)) {
+            }
+        }
+
+        Iterator begin() noexcept { return Iterator(nullptr, root_.get()); }
```
[the remark this answers is not in the document]
Comment (projedi) [reply runtime/src/mm/cpp/ObjectFactory.hpp]: Good point, thanks! Added special iterators for `Consumer`.

**Review by ilmat192 — APPROVED** [review]

**Pull request merged**